File: backend/app/services/pricing.py
from __future__ import annotations

from dataclasses import dataclass

DEFAULT_PRICING_MODEL = "gpt-5.2"


@dataclass(frozen=True)
class ModelPricing:
    input_per_million_usd: float
    output_per_million_usd: float


MODEL_PRICING: dict[str, ModelPricing] = {
    "gpt-5.2": ModelPricing(input_per_million_usd=1.75, output_per_million_usd=14.0),
    "gpt-5.2-chat-latest": ModelPricing(
        input_per_million_usd=1.75,
        output_per_million_usd=14.0,
    ),
    "gpt-5.2-codex": ModelPricing(input_per_million_usd=1.75, output_per_million_usd=14.0),
    "gpt-5.2-pro": ModelPricing(input_per_million_usd=21.0, output_per_million_usd=168.0),
    "gpt-5.1": ModelPricing(input_per_million_usd=1.25, output_per_million_usd=10.0),
    "gpt-5": ModelPricing(input_per_million_usd=1.25, output_per_million_usd=10.0),
    "gpt-5-mini": ModelPricing(input_per_million_usd=0.25, output_per_million_usd=2.0),
    "gpt-5-nano": ModelPricing(input_per_million_usd=0.05, output_per_million_usd=0.4),
    "o4-mini": ModelPricing(input_per_million_usd=1.1, output_per_million_usd=4.4),
    # Anthropic Claude models
    "claude-sonnet-4-6": ModelPricing(input_per_million_usd=3.0, output_per_million_usd=15.0),
    "claude-opus-4-6": ModelPricing(input_per_million_usd=15.0, output_per_million_usd=75.0),
    "claude-sonnet-4-20250514": ModelPricing(input_per_million_usd=3.0, output_per_million_usd=15.0),
    "claude-opus-4-20250514": ModelPricing(input_per_million_usd=15.0, output_per_million_usd=75.0),
}

DEFAULT_PRICING = MODEL_PRICING[DEFAULT_PRICING_MODEL]


def _strip_date_snapshot_suffix(model: str) -> str:
    import re

    return re.sub(r"-\d{4}-\d{2}-\d{2}$", "", model, flags=re.IGNORECASE)


def resolve_pricing_model(model: str) -> str:
    normalized = model.strip().lower()
    if normalized in MODEL_PRICING:
        return normalized

    without_date = _strip_date_snapshot_suffix(normalized)
    if without_date in MODEL_PRICING:
        return without_date

    if without_date.startswith("gpt-5.2-pro"):
        return "gpt-5.2-pro"
    if without_date.startswith("gpt-5.2-codex"):
        return "gpt-5.2-codex"
    if without_date.startswith("gpt-5.2-chat"):
        return "gpt-5.2-chat-latest"
    if without_date.startswith("gpt-5.2"):
        return "gpt-5.2"
    if without_date.startswith("gpt-5.1"):
        return "gpt-5.1"
    if without_date.startswith("gpt-5-mini"):
        return "gpt-5-mini"
    if without_date.startswith("gpt-5-nano"):
        return "gpt-5-nano"
    if without_date.startswith("gpt-5"):
        return "gpt-5"
    if without_date.startswith("o4-mini"):
        return "o4-mini"

    # Anthropic Claude models
    if without_date.startswith("claude-sonnet-4"):
        return "claude-sonnet-4-6"
    if without_date.startswith("claude-opus-4"):
        return "claude-opus-4-6"

    return DEFAULT_PRICING_MODEL


def estimate_text_token_cost_usd(
    model: str,
    input_tokens: int,
    output_tokens: int,
) -> tuple[float, str, ModelPricing]:
    pricing_model = resolve_pricing_model(model)
    pricing = MODEL_PRICING.get(pricing_model, DEFAULT_PRICING)
    input_cost = (max(input_tokens, 0) / 1_000_000) * pricing.input_per_million_usd
    output_cost = (max(output_tokens, 0) / 1_000_000) * pricing.output_per_million_usd
    return (input_cost + output_cost, pricing_model, pricing)

File: src/server/generate/pricing.ts
export interface ModelPricing {
  inputPerMillionUsd: number;
  outputPerMillionUsd: number;
}

const DEFAULT_PRICING_MODEL = "gpt-5.2";

const MODEL_PRICING: Record<string, ModelPricing> = {
  "gpt-5.2": { inputPerMillionUsd: 1.75, outputPerMillionUsd: 14.0 },
  "gpt-5.2-chat-latest": { inputPerMillionUsd: 1.75, outputPerMillionUsd: 14.0 },
  "gpt-5.2-codex": { inputPerMillionUsd: 1.75, outputPerMillionUsd: 14.0 },
  "gpt-5.2-pro": { inputPerMillionUsd: 21.0, outputPerMillionUsd: 168.0 },

  // Legacy fallbacks kept for older env values.
  "gpt-5.1": { inputPerMillionUsd: 1.25, outputPerMillionUsd: 10.0 },
  "gpt-5": { inputPerMillionUsd: 1.25, outputPerMillionUsd: 10.0 },
  "gpt-5-mini": { inputPerMillionUsd: 0.25, outputPerMillionUsd: 2.0 },
  "gpt-5-nano": { inputPerMillionUsd: 0.05, outputPerMillionUsd: 0.4 },
  "o4-mini": { inputPerMillionUsd: 1.1, outputPerMillionUsd: 4.4 },

  // Anthropic Claude models
  "claude-sonnet-4-6": { inputPerMillionUsd: 3.0, outputPerMillionUsd: 15.0 },
  "claude-opus-4-6": { inputPerMillionUsd: 15.0, outputPerMillionUsd: 75.0 },
};
const DEFAULT_PRICING = MODEL_PRICING[DEFAULT_PRICING_MODEL] as ModelPricing;

function normalizeModelId(model: string): string {
  return model.trim().toLowerCase();
}

function stripDateSnapshotSuffix(model: string): string {
  return model.replace(/-\d{4}-\d{2}-\d{2}$/i, "");
}

export function resolvePricingModel(model: string): string {
  const normalized = normalizeModelId(model);
  if (MODEL_PRICING[normalized]) return normalized;

  const withoutDate = stripDateSnapshotSuffix(normalized);
  if (MODEL_PRICING[withoutDate]) return withoutDate;

  if (withoutDate.startsWith("gpt-5.2-pro")) return "gpt-5.2-pro";
  if (withoutDate.startsWith("gpt-5.2-codex")) return "gpt-5.2-codex";
  if (withoutDate.startsWith("gpt-5.2-chat")) return "gpt-5.2-chat-latest";
  if (withoutDate.startsWith("gpt-5.2")) return "gpt-5.2";
  if (withoutDate.startsWith("gpt-5.1")) return "gpt-5.1";
  if (withoutDate.startsWith("gpt-5-mini")) return "gpt-5-mini";
  if (withoutDate.startsWith("gpt-5-nano")) return "gpt-5-nano";
  if (withoutDate.startsWith("gpt-5")) return "gpt-5";
  if (withoutDate.startsWith("o4-mini")) return "o4-mini";

  // Anthropic Claude models
  if (withoutDate.startsWith("claude-sonnet-4")) return "claude-sonnet-4-6";
  if (withoutDate.startsWith("claude-opus-4")) return "claude-opus-4-6";

  return DEFAULT_PRICING_MODEL;
}

export function estimateTextTokenCostUsd(
  model: string,
  inputTokens: number,
  outputTokens: number,
): { costUsd: number; pricingModel: string; pricing: ModelPricing } {
  const pricingModel = resolvePricingModel(model);
  const pricing = MODEL_PRICING[pricingModel] ?? DEFAULT_PRICING;
  const inputCost = (Math.max(inputTokens, 0) / 1_000_000) * pricing.inputPerMillionUsd;
  const outputCost =
    (Math.max(outputTokens, 0) / 1_000_000) * pricing.outputPerMillionUsd;

  return {
    costUsd: inputCost + outputCost,
    pricingModel,
    pricing,
  };
}
